crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,326 @@
+        format!("Invalid TODO capitalization: `{tag}` should be `TODO`")
+    }
+}
+
+/// ## What it does
+/// Checks that the colon after a "TODO" tag is followed by a space.
+///
+/// ## Why is this bad?
+/// Skipping the space after a colon leads to less readable code.
+///
+/// ## Example
+/// ```python
+/// # TODO(ruff):fix this
+/// ```
+///
+/// Use instead:
+/// ```python
+/// # TODO(ruff): fix this
+/// ```
+#[violation]
+pub struct MissingSpaceAfterColonInTodo;
+impl Violation for MissingSpaceAfterColonInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing space after colon in TODO")
+    }
+}
+
+// Capture groups correspond to checking:
+//      1. Tag (used to match against `TODO` in capitalization and spelling, e.g. ToDo and FIXME)
+//      2. Author exists (in parentheses) after tag, but before colon
+//      3. Colon exists after author
+//      4. Space exists after colon
+//      5. Text exists after space
+//
+// We can check if any of these exist in one regex. Capture groups that don't pick anything up
+// evaluate to `None` in Rust, so the capture group index will always correspond to its respective rule
+// whether the token has been found or not.
+//
+// Example:
+// ```python
+// # TODO(evanrittenhouse): This is a completely valid TODO
+// ```
+// will yield [Some("TODO") Some("evanrittenhouse"), Some(":"), Some(" "), Some("This is a completely valid TODO")], whereas
+// ```python
+// # ToDo this is completely wrong
+// ```
+// will yield [Some("ToDo"), None, None, Some(" "), Some("this is completely wrong")]. Note the
+// `Nones` for the colon and space checks.
+//
+// Note: Regexes taken from https://github.com/orsinium-labs/flake8-todos/blob/master/flake8_todos/_rules.py#L12.
+static TODO_REGEX: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(r"^#\s*(?P<tag>[tT][oO][dD][oO]|BUG|FIXME|XXX)(\(.*\))?(:)?( )?(.+)?$").unwrap()
+});
+
+// Issue code: TDO-003
+static ISSUE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r#"^#\s*[A-Z]{1,6}\-?\d+$"#).unwrap());
+// Issue code: 003
+static TICKET_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r#"^#\s*\d+$"#).unwrap());
+// Link to issue
+static LINK_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r#"^#\s*(http|https)://.*"#).unwrap());
+static NUM_CAPTURE_GROUPS: usize = 5usize;
+
+pub fn check_rules(tokens: &[LexResult]) -> Vec<Diagnostic> {
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+    let mut prev_token_is_todo = false;
+
+    for (start, token, end) in tokens.iter().flatten() {
+        let Tok::Comment(comment) = token else {
+            continue;
+        };
+
+        let diagnostics_ref = &mut diagnostics;
+        let range = Range::new(*start, *end);
+
+        // The previous token is a TODO, so let's check if this token is a link
+        if prev_token_is_todo {
+            if ISSUE_REGEX.is_match(comment)
+                || TICKET_REGEX.is_match(comment)
+                || LINK_REGEX.is_match(comment)
```

Comment:
Nit: It could be worth using [`RegexSet`](https://docs.rs/regex/latest/regex/struct.RegexSet.html) to avoid scanning the same text multiple times

Reply:
Great point! Didn't realize that exits - I'll change it in a couple
hours, out right now